### DISCUSSION TITLE
✨Tutorial: How to Use Permit2 with the Uniswap Universal Router | HackerNoon✨

### DISCUSSION THREAD

**Discussion by Elnaril** [Elnaril/uniswap-universal-router-decoder#27] [Announcements]
In this Python tutorial, you'll learn how to use the Permit2 contract with the Uniswap Universal Router, through its function **PERMIT2_PERMIT**. The point is to be able to give a Permit allowance to swap/sell ERC-20 tokens.

[Tutorial link](https://hackernoon.com/python-how-to-use-permit2-with-the-uniswap-universal-router)

The full tutorial code is [here](https://github.com/Elnaril/uniswap-universal-router-decoder/blob/master/tutorials/sell_tokens/sell_token.py).

**Comment by cryptofandorin** [comment]
Very interesting

**Comment by snarflakes** [comment]
I am unclear about what type of token approvals are assumed have been done to successfully use the encoder for uniswap v3 or v4 pool swaps.  Is it permit2 only? thanks!

**Reply by Elnaril** [reply]
@snarflakes , for V3 you can find examples in the integrations tests in addition to the tutorial. For V4, I'll add some integrations tests after the first stable release.
